### PR TITLE
Change Version switch output to finish with a newline

### LINKED ISSUE #9482: [Bug]: -version message doesn't have a line ending
### Issue Description

The -version switch displays the version number but doesn't include a line ending.

With Windows cmd and with PowerShell, this doesn't matter. The line is ended and the next prompt is shown on a new line.

But it shows up on macOS and Linux in a shell other than pwsh.

This is a bit of a nit but it is also a bit ugly and low effort to fix.

### Steps to Reproduce

On macOS or Linux in a standard shell run the following command:

```
dotnet msbuild --version -nologo
```

### Expected Behavior

Expected behavior is no artifacts and the next prompt is on a new line.

An example for the bash shell:

```
$ dotnet msbuild --version -nologo
17.8.3.51904
$
```

zsh should be the same, except the default zsh prompt is `%`.

### Actual Behavior

In bash on Ubuntu in WSL in Windows Terminal, the `$` command prompt for the next command is on the same line as the version output.

```
$ dotnet msbuild --version -nologo
17.8.3.51904$
```

Screenshot Linux bash

In zsh on macOS in Terminal, an inverse+bold `%` is displayed at the end of the version output (which is a zsh feature to show a partial line).

```
% dotnet msbuild --version -nologo
17.8.3.51904%
%
```

Screenshot macOS zsh

### Analysis

In the `MSBuildApp` class, the -version switch is mapped to a call to the `ShowVersion()` method.

https://github.com/dotnet/msbuild/blob/d8283be045c16620c3e2ae3bfe33367257c3b4f3/src/MSBuild/XMake.cs#L2497-L2502

https://github.com/dotnet/msbuild/blob/d8283be045c16620c3e2ae3bfe33367257c3b4f3/src/MSBuild/XMake.cs#L4488-L4494

Currently this is the only usage of `ShowVersion()`. The `Write` could be changed to a `Writeln`.

### Versions & Configurations

_No response_

### PR DESCRIPTION
**Pull request by jrdodds**
Fixes #9482

### Context
The -version switch doesn't terminate its output with a newline which some shells don't like.

### Changes Made
- Added a unit test.
- Changed `ShowVersion()`.

### Testing
Tested on Windows 10 and macOS 14.
Tested by running the full unit test suite and by manually running the `-version` switch under cmd (Windows), pwsh (Windows and macOS), and zsh (macOS).

### CODE REVIEW

**Review by baronfel — COMMENTED** [review]
Thank you!

**Review by rainersigwald — COMMENTED** [review]
This could, _conceivably_, break someone who has a script that's calling `msbuild -nologo -version | script_that_cant_handle_newline`, right?  I hate that I'm so conservative but I think I want it behind a changewave.

**Comment by jrdodds** [comment]
Yes, that's conceivable. And for someone who has written code like that and hits this change, the defect would be mysterious.

**Comment by rainersigwald** [comment]
That sounds fine to me.

**Comment by jrdodds** [comment]
Should this be in a change wave and should it be change wave 17.10?

**Comment by rainersigwald** [comment]
Yes and yes, please.

File: src/MSBuild.UnitTests/XMake_Tests.cs [hunk 1]
```diff
@@ -523,6 +523,87 @@ public void Help(string indicator)
...
+                },
+            };
+
+            process.Start();
+            process.WaitForExit();
+            process.ExitCode.ShouldBe(0);
+
+            string output = process.StandardOutput.ReadToEnd();
+            // Change Version switch output to finish with a newline https://github.com/dotnet/msbuild/pull/9485
+            if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_10))
```
Comment (f-alizada) [inline src/MSBuild.UnitTests/XMake_Tests.cs:556]: May I clarify if there is a case for this test case, when the feature will not be available => hence the output will not contain the new line? Since the feature is opt-out, new line will be written to output unless change wave is configured, which is tested in separate test: VersionSwitchDisableChangeWave.

**Review by f-alizada — COMMENTED** [review]
Thank you for the PR! 
I left one comment with question about the verification of two scenarios in one test, overall looks good :)

File: src/MSBuild.UnitTests/XMake_Tests.cs (hunk 1 above)
Comment (f-alizada) [reply src/MSBuild.UnitTests/XMake_Tests.cs:556]: Thank you for the explanation, basically the test covers both scenarios relying on the setup of local environment.
I would suggest not to rely on the local environment in unit tests, and cover in this test success path of the ShowVersion method when there is no ChangeWaves enabled. As you suggested to remove env var if exists and proceed to the test :)

Comment (rainersigwald) [reply src/MSBuild.UnitTests/XMake_Tests.cs:556]: Our `TestEnvironment` class should clear the changewave state when it starts up--but today it does so only when it _cleans_ up, which is awkward and doesn't provide a strong guarantee.

Comment (jrdodds) [reply src/MSBuild.UnitTests/XMake_Tests.cs]: Committed changes to use TestEnvironment. PR builds passed.

File: eng/Versions.props [hunk 2]
```diff
@@ -2,7 +2,7 @@
 <!-- Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the MIT license. See License.txt in the project root for full license information. -->
 <Project>
   <PropertyGroup>
-    <VersionPrefix>17.9.0</VersionPrefix>
+    <VersionPrefix>17.10.0</VersionPrefix>
```
Comment (f-alizada) [inline eng/Versions.props:5]: It seems thtat this change comes from the merging main into feature branch. 
since the target branch is vs17.9 will it be possible to merge vs17.9 instead of main branch?

**Review by f-alizada — COMMENTED** [review]
Thank you for the updates! Left one comment regarding the base branch to be merged

File: eng/Versions.props (hunk 2 above)
Comment (jrdodds) [reply eng/Versions.props]: reverted the merge from main

File: eng/Version.Details.xml [hunk 3]
```diff
@@ -72,9 +72,9 @@
       <Uri>https://github.com/nuget/nuget.client</Uri>
       <Sha>707c46e558b2b027d7ae942028c369e26545f10a</Sha>
     </Dependency>
-    <Dependency Name="Microsoft.Net.Compilers.Toolset" Version="4.9.0-3.23602.1">
+    <Dependency Name="Microsoft.Net.Compilers.Toolset" Version="4.9.0-2.23575.1">
```
Comment (f-alizada) [inline eng/Version.Details.xml:75]: May I clarify if this is something required for or in scope of this PR? 
Usually those updates are done automatically for example: https://github.com/dotnet/msbuild/pull/9517/files

File: eng/Versions.props (hunk 2 above)
Comment (f-alizada) [reply eng/Versions.props]: Thank you for updates!

File: eng/Version.Details.xml (hunk 3 above)
Comment (f-alizada) [reply eng/Version.Details.xml:75]: Thank you @jrdodds for clarification! 
I just checked this file the main branch and the vs17.9 branch
main: https://github.com/dotnet/msbuild/blob/main/eng/Version.Details.xml
vs17.9: https://github.com/dotnet/msbuild/blob/vs17.9/eng/Version.Details.xml
And it seems that the version in both of the branches is "4.9.0-3.23602.1", maybe the main branch was not up to date when it was merged. I can see that the version got updated last week in this PR: https://github.com/dotnet/msbuild/commit/0e33df147b4bf7e5d980a325953fddd73975cf32

The same pattern could be seen in the https://github.com/dotnet/msbuild/commit/2f3d37672a69142a13a62856b09034a915bedc70 that those changes are being reverted in this PR. 

Basically as part of the PR it is expected only files that are related to the scope of the problem to be changed.
Sorry for the confusions, it is important for us to have as granular changes as possible in order to identify possible regressions or bugs earlier:) 

Please correct me if there are mistakes in my steps of verification, I will be glad to double check it

Comment (jrdodds) [reply eng/Version.Details.xml:75]: Are you asking that PRs not be kept in sync with the target branch? I haven't merged from vs17.9 which you requested earlier. Are you now requesting that not be done?

Comment (rainersigwald) [reply eng/Version.Details.xml:75]: We're in a weird state right now where changes are flowing into 17.9 and main, so reverting a merge from main is still resulting in an overall change of some unrelated things in this PR for 17.9. I think the best things to do would be to interactive-rebase onto 17.9. I can do that for you if you'd like.

Comment (jrdodds) [reply eng/Version.Details.xml:75]: If it is not a problem, please go ahead and rebase the PR. (I can't give the PR much of my attention at present.) Thanks

Comment (rainersigwald) [reply eng/Version.Details.xml]: Not a problem, done!

**Review by rainersigwald — APPROVED** [review]

**Review by f-alizada — APPROVED** [review]

File: eng/Version.Details.xml (hunk 3 above)
Comment (f-alizada) [reply eng/Version.Details.xml]: Resolving the conversation. Thank you @rainersigwald for the updates
